src/main/java/org/powerflows/dmn/engine/evaluator/expression/provider/MvelExpressionEvaluationProvider.java

```diff
@@ -88,38 +94,71 @@ public Serializable evaluateOutputEntry(final OutputEntry outputEntry, final Eva
     }
 
     Serializable evaluate(final InputEntry inputEntry, final EvaluationContext evaluationContext) {
-        final VariableResolverFactory mapVariableResolverFactory = new MapVariableResolverFactory();
-
-        fillVariables(evaluationContext, mapVariableResolverFactory);
+        final VariableResolverFactory mapVariableResolverFactory = fillVariables(evaluationContext);
         mapVariableResolverFactory.createVariable(inputEntry.getNameAlias(), evaluationContext.get(inputEntry.getName()));
 
         return evaluate(inputEntry.getExpression(), mapVariableResolverFactory);
     }
 
     final Serializable evaluate(final Expression expression, final EvaluationContext evaluationContext) {
-        final VariableResolverFactory mapVariableResolverFactory = new MapVariableResolverFactory();
-
-        fillVariables(evaluationContext, mapVariableResolverFactory);
-        return evaluate(expression, mapVariableResolverFactory);
+        return evaluate(expression, fillVariables(evaluationContext));
     }
 
-    final void fillVariables(final EvaluationContext evaluationContext, final VariableResolverFactory mapVariableResolverFactory) {
-        evaluationContext
-                .getAll()
-                .keySet()
-                .forEach(variableName -> mapVariableResolverFactory.createVariable(variableName, evaluationContext.get(variableName)));
+    final VariableResolverFactory fillVariables(final EvaluationContext evaluationContext) {
+        return new FastCachingMapVariableResolverFactory(evaluationContext.getAll());
     }
 
     final Serializable evaluate(final Expression expression, final VariableResolverFactory variableResolverFactory) {
         final Serializable result;
         variableResolverFactory.setNextFactory(functionResolverFactory);
 
         try {
-            result = (Serializable) MVEL.eval((String) expression.getValue(), variableResolverFactory);
+            final String expressionString = (String) expression.getValue();
+            final Object compiledExpression = expressionCache.computeIfAbsent(expressionString, MVEL::compileExpression);
+            result = (Serializable) MVEL.executeExpression(compiledExpression, variableResolverFactory);
         } catch (Exception e) {
             throw new ExpressionEvaluationException("Can not evaluate feel expression '" + expression.getValue() + "'", e);
         }
 
         return result;
     }
+
+    /**
+     * {@link CachingMapVariableResolverFactory} uses exception driven programming instead of null checks.
+     */
+    private static class FastCachingMapVariableResolverFactory extends CachingMapVariableResolverFactory {
```

Comment:
Can you consider to move to separated file? Meybe would be desirable to create mvel package?